src/QsCompiler/Compiler/RewriteSteps/FunctorGeneration.cs

```diff
@@ -0,0 +1,62 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.Quantum.QsCompiler.SyntaxTree;
+
+
+namespace Microsoft.Quantum.QsCompiler.BuiltInRewriteSteps
+{
+    internal class FunctorGeneration : IRewriteStep
+    {
+        public string Name { get; }
+        public int Priority { get; }
+        public IDictionary<string, string> AssemblyConstants { get; }
+        public IEnumerable<IRewriteStep.Diagnostic> GeneratedDiagnostics { get; }
+
+        public bool ImplementsTransformation { get; }
+        public bool ImplementsPreconditionVerification { get; }
+        public bool ImplementsPostconditionVerification { get; }
+
+        public FunctorGeneration()
+        {
+            Name = "Functor Generation";
+            Priority = 10; // Not used for built-in transformations like this
+            AssemblyConstants = new Dictionary<string, string>();
+            ImplementsTransformation = true;
+            ImplementsPreconditionVerification = true;
+            ImplementsPostconditionVerification = false;
+        }
+
+        public bool Transformation(QsCompilation compilation, out QsCompilation transformed)
+        {
+            return CodeGeneration.GenerateFunctorSpecializations(compilation, out transformed);
+        }
+
+        public bool PreconditionVerification(QsCompilation compilation)
+        {
+            var requiredNamespace = compilation.Namespaces
+                .FirstOrDefault(ns => ns.Name.Equals(BuiltIn.CoreNamespace));
+
+            if (requiredNamespace == null)
+            {
+                return false;
+            }
+
+            var providedOperations = new QsNamespace[] { requiredNamespace }.Callables().Select(c => c.FullName);
+            var requiredBuiltIns = new List<QsQualifiedName>()
+            {
+                BuiltIn.Length.FullName,
+                BuiltIn.RangeReverse.FullName
+            };
+
+            return requiredBuiltIns.All(builtIn => providedOperations.Any(provided => provided.Equals(builtIn)));
```

Comment:
It's two entries and not a whole lot of namespaces, so it doesn't really matter, but as a general setup, I'd expect that making providedOperations something like an ImmutableSet and leveraging set operations should be more efficient.

Reply:
Hmm, I am skeptical that using set operations would be more efficient than using the Any method in general for these kinds of checks. Do you have any resources that you could post?